File: dart/geiger_api/lib/src/communication/communication_helper.dart
library geiger_api;

import '../../geiger_api.dart';

/// Interface to denote a message filter.
typedef MessageFilter = bool Function(Message msg);

class Listener with PluginListener {
  final GeigerApi api;
  final Object obj = Object();
  Message tmsg;
  Message? msg;
  List<MessageType> responseTypes;

  Listener(this.api, this.tmsg, this.responseTypes) {
    api.registerListener(responseTypes, this);
  }

  @override
  void pluginEvent(GeigerUrl? url, Message msg) {
    if (this.msg == null &&
        tmsg.requestId == msg.requestId &&
        tmsg.targetId == msg.sourceId &&
        tmsg.sourceId == msg.targetId) {
      this.msg = msg;
    }
  }

  void dispose() {
    api.deregisterListener(responseTypes, this);
  }

  Future<Message> waitForResult(int timeout) async {
    var startTime = DateTime.now().millisecondsSinceEpoch;
    while ((msg == null) &&
        ((timeout < 0) ||
            ((DateTime.now().millisecondsSinceEpoch - startTime) < timeout))) {
      await Future.delayed(const Duration(milliseconds: 100));
    }
    var message = msg;
    if (message == null) {
      throw CommunicationException('timeout reached while waiting for reply');
    }
    return message;
  }
}

/// A helper class for sending and waiting on [Message]s.
class CommunicationHelper {
  /// Sends [msg] and waits for the first message matching the provided [filter].
  ///
  /// Will communication using the provided [api] and waits maximum [timeout]
  /// milliseconds. Specify `-1` to remove any time limit.
  ///
  /// Throws [CommunicationException] if communication with master fails
  static Future<Message> sendAndWait(GeigerApi api, Message msg,
      {int timeout = 30000,
      List<MessageType> responseTypes = const [
        MessageType.comapiSuccess,
        MessageType.comapiError
      ]}) async {
    var l = Listener(api, msg, responseTypes);
    await api.sendMessage(msg);
    var result = await l.waitForResult(timeout);
    l.dispose();
    return result;
  }
}

File: java/GeigerApiPluginConnector/src/main/java/eu/cybergeiger/api/communication/CommunicationHelper.java
package eu.cybergeiger.api.communication;

import eu.cybergeiger.api.GeigerApi;
import eu.cybergeiger.api.message.Message;
import eu.cybergeiger.api.message.MessageType;
import eu.cybergeiger.api.plugin.PluginListener;

import java.io.IOException;
import java.util.Objects;
import java.util.concurrent.TimeoutException;

/**
 * A helper class for sending and waiting on Messages.
 * TODO should this only be used for Testing?
 */
public class CommunicationHelper {
  private static class Listener implements PluginListener {
    private final GeigerApi api;
    private final Message requestMessage;
    private final MessageType[] responseTypes;
    private Message responseMessage = null;
    private final Object receivedResponse = new Object();


    public Listener(GeigerApi api, Message requestMessage, MessageType[] responseTypes) {
      this.api = api;
      this.requestMessage = requestMessage;
      this.responseTypes = responseTypes;
      this.api.registerListener(responseTypes, this);
    }

    @Override
    public void pluginEvent(Message message) {
      if (responseMessage != null ||
        !Objects.equals(requestMessage.getRequestId(), message.getRequestId()) ||
        !Objects.equals(requestMessage.getTargetId(), message.getSourceId()) ||
        !Objects.equals(requestMessage.getSourceId(), message.getTargetId()))
        return;
      synchronized (receivedResponse) {
        responseMessage = message;
        receivedResponse.notifyAll();
      }
    }

    public void dispose() {
      api.deregisterListener(responseTypes, this);
    }

    public Message waitForResult(long timeoutMillis) throws TimeoutException, InterruptedException {
      if (responseMessage == null) {
        synchronized (receivedResponse) {
          receivedResponse.wait(timeoutMillis);
        }
      }
      if (responseMessage == null) {
        throw new TimeoutException("Timeout reached while waiting for reply.");
      }
      return responseMessage;
    }
  }


  /**
   * <p>Sends a message and waits for the first
   * returning message of a specific type with the same requestId.</p>
   *
   * @param api     GeigerAPI to communicate over.
   * @param message Message to send.
   * @return The response Message.
   */
  public static Message sendAndWait(GeigerApi api, Message message)
    throws InterruptedException, TimeoutException, IOException {
    return sendAndWait(api, message, new MessageType[]{MessageType.COMAPI_SUCCESS, MessageType.COMAPI_ERROR});
  }

  /**
   * <p>Sends a message and waits for the first
   * returning message of a specific type with the same requestId.</p>
   *
   * @param api           GeigerAPI to communicate over.
   * @param message       Message to send.
   * @param responseTypes Possible message types of response message.
   * @return The response Message.
   */
  public static Message sendAndWait(GeigerApi api, Message message, MessageType[] responseTypes)
    throws InterruptedException, TimeoutException, IOException {
    return sendAndWait(api, message, responseTypes, 30000);
  }

  /**
   * <p>Sends a message and waits for the first
   * returning message of a specific type with the same requestId.</p>
   *
   * @param api           GeigerAPI to communicate over.
   * @param message       Message to send.
   * @param responseTypes Possible message types of response message.
   * @param timeoutMillis Timeout in milliseconds.
   * @return The response Message.
   */
  public static Message sendAndWait(GeigerApi api, Message message,
                                    MessageType[] responseTypes, long timeoutMillis)
    throws InterruptedException, TimeoutException, IOException {
    Listener listener = new Listener(api, message, responseTypes);
    api.sendMessage(message);
    Message result = listener.waitForResult(timeoutMillis);
    listener.dispose();
    return result;
  }
}
